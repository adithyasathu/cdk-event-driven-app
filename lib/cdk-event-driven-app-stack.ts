import { Stack, StackProps, CfnOutput } from 'aws-cdk-lib';
import { NodejsFunction} from "aws-cdk-lib/aws-lambda-nodejs";

import { Construct } from 'constructs';
import { RestApi, LambdaIntegration, MethodLoggingLevel } from 'aws-cdk-lib/aws-apigateway';
import { Runtime, Tracing } from 'aws-cdk-lib/aws-lambda';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';

export class CdkEventDrivenAppStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const esBuildSettings = {
      sourceMap: true,
      minify: true
    }

    const envVariables = {
      AWS_ACCOUNT_ID: Stack.of(this).account,
  };

    const functionSettings = {
      handler: "index.handler",
      runtime: Runtime.NODEJS_20_X,
      memorySize: 256,
      environment: {
        ...envVariables
      },
      logRetention: RetentionDays.ONE_WEEK,
      tracing: Tracing.ACTIVE,
      bundling: esBuildSettings
    }

    // Define a new Lambda function
    const healthLambda = new NodejsFunction(this, 'HealthHandler', {
      ...functionSettings,
      entry: 'src/health/index.ts',
      
    });

    // Define a new API Gateway REST API resource
    const api = new RestApi(this, 'Platform-X-API', {
      restApiName: 'API Gateway for the Platform X',
      description: 'This service provides backend APIs for Platform X',
    });

    // Define a new GET method for the /health resource
    const getHealthIntegration = new LambdaIntegration(healthLambda);
    api.root.addResource('health').addMethod('GET', getHealthIntegration);

    new CfnOutput(this, 'ApiUrl', {
      value: `${api.url}health` 
    })
  }
}
