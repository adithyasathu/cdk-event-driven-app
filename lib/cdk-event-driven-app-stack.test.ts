import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { CdkEventDrivenAppStack } from "./cdk-event-driven-app-stack";

describe("API Gateway Created", () => {
  let app: App;
  let stack: CdkEventDrivenAppStack;
  let template: Template;
  beforeAll(() => {
    app = new App();
    // WHEN
    stack = new CdkEventDrivenAppStack(app, "MyTestStack");
    // THEN

    template = Template.fromStack(stack);
  });

  test("API Gateway Created", () => {
    template.resourceCountIs("AWS::ApiGateway::RestApi", 1);
  });

  test("API Gateway has health endpoint", () => {
    template.hasResourceProperties("AWS::ApiGateway::Resource", {
      PathPart: "health",
    });
  });

  test("matches the snapshot", () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});
