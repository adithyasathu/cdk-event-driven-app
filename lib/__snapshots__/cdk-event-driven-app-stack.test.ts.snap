// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API Gateway Created matches the snapshot 1`] = `
{
  "Outputs": {
    "ApiUrl": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "PlatformXAPI88A6D63D",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "PlatformXAPIDeploymentStageprodC8D79539",
            },
            "/health",
          ],
        ],
      },
    },
    "PlatformXAPIEndpointE99B8A65": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "PlatformXAPI88A6D63D",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "PlatformXAPIDeploymentStageprodC8D79539",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "HealthHandler164742B2": {
      "DependsOn": [
        "HealthHandlerServiceRoleDefaultPolicy65C62A94",
        "HealthHandlerServiceRole2FE1C1C8",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "eb0b2e9e25bc1d48b0ab861f369d7848939d1604833e1c534358ebc2069ef3df.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_ACCOUNT_ID": {
              "Ref": "AWS::AccountId",
            },
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "HealthHandlerServiceRole2FE1C1C8",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "HealthHandlerLogRetentionB4F1C426": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "HealthHandler164742B2",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "HealthHandlerServiceRole2FE1C1C8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "HealthHandlerServiceRoleDefaultPolicy65C62A94": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "HealthHandlerServiceRoleDefaultPolicy65C62A94",
        "Roles": [
          {
            "Ref": "HealthHandlerServiceRole2FE1C1C8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "4e26bf2d0a26f2097fb2b261f22bb51e3f6b4b52635777b1e54edbd8e2d58c35.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PlatformXAPI88A6D63D": {
      "Properties": {
        "Description": "This service provides backend APIs for Platform X",
        "Name": "API Gateway for the Platform X",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "PlatformXAPIAccount3FF01F4F": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "PlatformXAPI88A6D63D",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "PlatformXAPICloudWatchRole33C9C75D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "PlatformXAPICloudWatchRole33C9C75D": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "PlatformXAPIDeploymentC725E86F181e18e68d74e162bfd4bcf78862f2ce": {
      "DependsOn": [
        "PlatformXAPIhealthGET5D562A5F",
        "PlatformXAPIhealth5D334476",
      ],
      "Properties": {
        "Description": "This service provides backend APIs for Platform X",
        "RestApiId": {
          "Ref": "PlatformXAPI88A6D63D",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "PlatformXAPIDeploymentStageprodC8D79539": {
      "DependsOn": [
        "PlatformXAPIAccount3FF01F4F",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "PlatformXAPIDeploymentC725E86F181e18e68d74e162bfd4bcf78862f2ce",
        },
        "RestApiId": {
          "Ref": "PlatformXAPI88A6D63D",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "PlatformXAPIhealth5D334476": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "PlatformXAPI88A6D63D",
            "RootResourceId",
          ],
        },
        "PathPart": "health",
        "RestApiId": {
          "Ref": "PlatformXAPI88A6D63D",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "PlatformXAPIhealthGET5D562A5F": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "HealthHandler164742B2",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "PlatformXAPIhealth5D334476",
        },
        "RestApiId": {
          "Ref": "PlatformXAPI88A6D63D",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "PlatformXAPIhealthGETApiPermissionMyTestStackPlatformXAPI8A5366C8GEThealth163A0CD9": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "HealthHandler164742B2",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "PlatformXAPI88A6D63D",
              },
              "/",
              {
                "Ref": "PlatformXAPIDeploymentStageprodC8D79539",
              },
              "/GET/health",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "PlatformXAPIhealthGETApiPermissionTestMyTestStackPlatformXAPI8A5366C8GEThealthD0C78801": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "HealthHandler164742B2",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "PlatformXAPI88A6D63D",
              },
              "/test-invoke-stage/GET/health",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
